x/incentives/keeper/distribute_test.go

```diff
@@ -171,98 +178,238 @@ func (suite *KeeperTestSuite) TestSyntheticDistribute() {
-	suite.Require().Equal(coins, sdk.Coins(nil))
+	for _, tc := range tests {
+		suite.SetupTest()
 
-	// setup a non perpetual lock and gauge
-	_, gaugeID, gaugeCoins, startTime := suite.SetupLockAndGauge(false)
+		incentivesKeeper := suite.App.IncentivesKeeper
 
-	// check that the sum of coins yet to be distributed is equal to the newly created gaugeCoins
-	coins = suite.App.IncentivesKeeper.GetModuleToDistributeCoins(suite.Ctx)
-	suite.Require().Equal(coins, gaugeCoins)
+		// check that the sum of coins yet to be distributed is nil
+		moduleToDistributeCoins := incentivesKeeper.GetModuleToDistributeCoins(suite.Ctx)
+		suite.Equal(moduleToDistributeCoins, sdk.Coins(nil))
+		// gauges used in this test
+		gauges := []*types.Gauge{}
 
-	// add coins to the previous gauge and check that the sum of coins yet to be distributed includes these new coins
-	addCoins := sdk.Coins{sdk.NewInt64Coin("stake", 200)}
-	suite.AddToGauge(addCoins, gaugeID)
-	coins = suite.App.IncentivesKeeper.GetModuleToDistributeCoins(suite.Ctx)
-	suite.Require().Equal(coins, gaugeCoins.Add(addCoins...))
+		expModuleToDistributeCoins := sdk.Coins{}
+		// coins that will be distributed after the first epoch for each gauge
+		expDistributedCoinsFirstEpoch := map[uint64]sdk.Coins{}
 
-	// create a new gauge
-	// check that the sum of coins yet to be distributed is equal to the gauge1 and gauge2 coins combined
-	_, _, gaugeCoins2, _ := suite.SetupNewGauge(false, sdk.Coins{sdk.NewInt64Coin("stake", 1000)})
-	coins = suite.App.IncentivesKeeper.GetModuleToDistributeCoins(suite.Ctx)
-	suite.Require().Equal(coins, gaugeCoins.Add(addCoins...).Add(gaugeCoins2...))
+		suite.LockTokens(lockOwner, sdk.Coins{sdk.NewInt64Coin("lptoken", 10)}, time.Second)
 
-	// move all created gauges from upcoming to active
-	suite.Ctx = suite.Ctx.WithBlockTime(startTime)
-	gauge, err := suite.App.IncentivesKeeper.GetGaugeByID(suite.Ctx, gaugeID)
-	suite.Require().NoError(err)
-	err = suite.App.IncentivesKeeper.MoveUpcomingGaugeToActiveGauge(suite.Ctx, *gauge)
-	suite.Require().NoError(err)
+		// create new gauge using initialGaugeCoins
+		for _, coins := range tc.initialGaugeCoins {
+			_, gauge, gaugeCoins, _ := suite.SetupNewGauge(false, coins)
+			suite.Equal(coins, gaugeCoins, "test %v", tc.name)
 
-	// distribute coins to stakers
-	distrCoins, err := suite.App.IncentivesKeeper.Distribute(suite.Ctx, []types.Gauge{*gauge})
-	suite.Require().NoError(err)
-	suite.Require().Equal(distrCoins, sdk.Coins{sdk.NewInt64Coin("stake", 105)})
+			gauges = append(gauges, gauge)
+
+			moduleToDistributeCoins = incentivesKeeper.GetModuleToDistributeCoins(suite.Ctx)
+			expModuleToDistributeCoins = expModuleToDistributeCoins.Add(gaugeCoins...)
+			suite.Equal(expModuleToDistributeCoins, moduleToDistributeCoins, "test %v", tc.name)
+
+			// div by 2 since all gauge in this tests distribute over 2 epochs
+			expDistributedCoinsFirstEpoch[gauge.Id] = DivCoin(coins, 2)
```

Comment:
Curious, what happens to roundings when perfect division is not possible? Ex) three epochs with gauge of 1000 stake coins. 

Can we add test case field with epochs, add tests with cases that can't be perfectly divided or test cases with epoch 0?

Reply:
yeah, you're right, how about we have odd gauge coins and 2 epochs ?

Reply:
Although it shouldn't matter, maybe we also add odd epochs and even gauge just to cover all our bases (in case of future refactor of distribution logic)

Reply:
test cases added, also I've changed `SetupNewGaugeWithDenom` and  `SetupNewGauge` so that they takes in additional param `numEpochsPaidOver`